src/java.base/share/classes/jdk/internal/template/TemplateSupport.java

```diff
@@ -0,0 +1,207 @@
+     */
+    private final List<String> fragments;
+
+    /**
+     * Static final processor that triggered the BSM generation.
+     */
+    private final ValidatingProcessor<?, ?> processor;
+
+    private static final JavaUtilCollectionAccess JUCA = SharedSecrets.getJavaUtilCollectionAccess();
+
+    /**
+     * Initialize {@link MethodHandle MethodHandles}.
+     */
+    static {
+        try {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+
+            MethodType mt = MethodType.methodType(Object.class,
+                    List.class, ValidatingProcessor.class, Object[].class);
+            DEFAULT_PROCESS_MH = lookup.findStatic(TemplateSupport.class, "defaultProcess", mt);
+        } catch (ReflectiveOperationException ex) {
+            throw new AssertionError("string bootstrap fail", ex);
+        }
+    }
+
+    /**
+     * Constructor.
+     *
+     * @param lookup    method lookup
+     * @param name      method name
+     * @param type      method type
+     * @param fragments fragments from string template
+     * @param processor static final processor
+     */
+    TemplateSupport(MethodHandles.Lookup lookup, String name, MethodType type,
+                    List<String> fragments,
+                    ValidatingProcessor<?, ?> processor) {
+        this.lookup = lookup;
+        this.name = name;
+        this.type = type;
+        this.fragments = fragments;
+        this.processor = processor;
+
+    }
+
+    /**
+     * Templated string bootstrap method.
+     *
+     * @param lookup          method lookup
+     * @param name            method name
+     * @param type            method type
+     * @param processorGetter {@link MethodHandle} to get static final processor
+     * @param fragments       fragments from string template
+     * @return {@link CallSite} to handle templated string processing
+     * @throws NullPointerException if any of the arguments is null
+     * @throws Throwable            if linkage fails
+     */
+    public static CallSite stringTemplateBSM(
+            MethodHandles.Lookup lookup,
+            String name,
+            MethodType type,
+            MethodHandle processorGetter,
+            String... fragments) throws Throwable {
+        Objects.requireNonNull(lookup, "lookup is null");
+        Objects.requireNonNull(name, "name is null");
+        Objects.requireNonNull(type, "type is null");
+        Objects.requireNonNull(processorGetter, "processorGetter is null");
+        Objects.requireNonNull(fragments, "fragments is null");
+
+        MethodType processorGetterType = MethodType.methodType(ValidatingProcessor.class);
+        ValidatingProcessor<?, ?> processor =
+                (ValidatingProcessor<?, ?>)processorGetter.asType(processorGetterType).invokeExact();
+        TemplateSupport support = new TemplateSupport(lookup, name, type, List.of(fragments), processor);
+
+        return support.processWithProcessor();
+    }
```

Comment:
Thinking about this protocol some more: this seems to depend on the processor being a constant, which precludes specialized linking of call sites where the processor is a dynamic argument.

The returned method handle/call site could maybe be changed to take the processor instance as a leading argument instead. The processor can then be used to pass additional arguments to the processing code. (and, AFAICS, that will also allow using calls to this BSM as the sole translation strategy for every type of processor, which seems more robust if types are changed later on to (un-)implement `ProcessorLinkage`)

The `linkage` method could instead be a `static` method, which is somehow tied to the type of the processor. Since it's currently a sealed interface you could have a mapping from each implementer to the `linkage` method for the types you care about (only `FormatProcessor` atm). If that is to be opened up to public extension in the future, something like type classes would be needed I think, so that the runtime can reliably map from the processor type to the static `linkage` method.

WDYT?

Reply:
(FWIW, I don't see this as prohibitive for this PR to go ahead, but maybe something to consider before finalizing the feature)

Reply:
Yes, the current code only supports processors with no fields, which is Okay given that the interface  `ProcessorLinkage` is sealed.

If/When the processor linkage API is open for everyone as you said the design will have to be changed.
But going the other way, creates a general mechanism is painful because either you have an adhoc way to associate  the processor to a BSM-like method or you have to delay until runtime the linkage using a monomorphic/polymorphic cache (I know Jim has tested that before back-pedaling to a simpler more constrained good enough solution).